// generate-license-file/src/lib/models/license.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.License = void 0;
const string_utils_1 = require("../utils/string.utils");
const BULLET = " - ";
const PREFIX = "The following npm package may be included in this product:";
const PREFIX_PLURAL = "The following npm packages may be included in this product:";
const MIDFIX = "This package contains the following license and notice below:";
const MIDFIX_PLURAL = "These packages each contain the following license and notice below:";
class License {
    constructor(content, dependencies) {
        this.content = content;
        this.dependencies = dependencies;
    }
    format(lineEnding) {
        const formattedText = this.prefix(lineEnding) +
            this.formatDependencies(lineEnding) +
            this.midfix(lineEnding) +
            (0, string_utils_1.prepareContentForOutput)(this.content.trim(), lineEnding);
        return formattedText;
    }
    prefix(EOL) {
        if (this.dependencies.length === 1) {
            return PREFIX + EOL + EOL;
        }
        return PREFIX_PLURAL + EOL + EOL;
    }
    formatDependencies(EOL) {
        const formattedText = this.dependencies
            .sort((a, b) => a.localeCompare(b))
            .map(dependency => {
            return BULLET + dependency + EOL;
        });
        return formattedText.join("");
    }
    midfix(EOL) {
        if (this.dependencies.length === 1) {
            return EOL + MIDFIX + EOL + EOL;
        }
        return EOL + MIDFIX_PLURAL + EOL + EOL;
    }
}
exports.License = License;
